lib/ProductOpener/Packaging.pm

```diff
@@ -341,6 +345,283 @@ sub guess_language_of_packaging_text ($text, $potential_lcs_ref) {
+
+	# If the shape is "capsule" and the product is in category "en:coffees", mark the shape as a "coffee capsule"
+	if (    (defined $packaging_ref->{"shape"})
+		and ($packaging_ref->{"shape"} eq "en:capsule")
+		and (has_tag($product_ref, "categories", "en:coffees")))
+	{
+		$packaging_ref->{"shape"} = "en:coffee-capsule";
+	}
+
+	# If we have a shape without a material, check if there is a default material for the shape
+	# e.g. "en:Bubble wrap" has the property packaging_materials:en: en:plastic
+	if ((defined $packaging_ref->{"shape"}) and (not defined $packaging_ref->{"material"})) {
+		my $material = get_inherited_property("packaging_shapes", $packaging_ref->{"shape"}, "packaging_materials:en");
+		if (defined $material) {
+			$packaging_ref->{"material"} = $material;
+		}
+	}
+
+	# If we have a material without a shape, check if there is a default shape for the material
+	# e.g. "en:tetra-pak" has the shape "en:brick"
+	if ((defined $packaging_ref->{"material"}) and (not defined $packaging_ref->{"shape"})) {
+		my $shape = get_inherited_property("packaging_materials", $packaging_ref->{"material"}, "packaging_shapes:en");
+		if (defined $shape) {
+			$packaging_ref->{"shape"} = $shape;
+		}
+	}
+	return;
+}
+
+=head2 add_or_combine_packaging_component_data($product_ref, $packaging_ref, $response_ref)
+
+This function adds the data for a packaging component to the packagings data structure,
+or if the packaging component data is compatible with an existing component
+of the packagings structure, the data is combined.
+
+=head3 Parameters
+
+=head4 $product_ref product
+
+=head4 $packaging_ref packaging component
+
+=head4 $response_ref API response object reference
+
+The API response object is used to return warnings and errors to the caller.
+If a warning or error is found (e.g. an unrecognized input), it is returned in
+the "warnings" or "errors" array of the response object.
+
+=cut
+
+sub add_or_combine_packaging_component_data ($product_ref, $packaging_ref, $response_ref) {
+
+	$log->debug("add_or_combine_packaging_component_data - start", {packaging_ref => $packaging_ref})
+		if $log->is_debug();
+
+	# Non empty packaging?
+	if ((scalar keys %$packaging_ref) > 0) {
+
+		# If we have an existing packaging that can correspond, augment it
+		# otherwise, add one
+
+		my $matching_packaging_ref;
+
+		foreach my $existing_packaging_ref (@{$product_ref->{packagings}}) {
+
+			my $match = 1;
+
+			foreach my $property (sort keys %$packaging_ref) {
+
+				# If the existing packaging does not have a property, it can match
+				if (not defined $existing_packaging_ref->{$property}) {
+					next;
+				}
+
+				my $tagtype = $packaging_taxonomies{$property};
+
+				# $tagtype can be shape / material / recycling, or undef if the property is something else (e.g. a number of packagings)
+				if (not defined $tagtype) {
+					# If there is an existing value for the property,
+					# check if it is the same
```

Comment:
why so ?
So if there was 1 plastic bag, and I fix for 2 plastic bag, this will indeed create a new entry ?

Reply:
If you fix something, then you use packagings and send the whole structure, not packagings_add.

packagings_add is used when the caller does not know the whole structure, but parts of it. e.g. Robotoff sees a PET logo, so it sends something like packagings_add -> { material : "pet" }